Exception switching projects

On Mac OSX, PyCharm 2017.1.4 with `intellij-rust` 0.1.0.2060.

Switching from an open Rust project (`rust-belt`) to another caused an exception:

```
Already disposed: Project (Disposed) rust-belt
java.lang.AssertionError: Already disposed: Project (Disposed) rust-belt
	at com.intellij.openapi.components.impl.ComponentManagerImpl.getPicoContainer(ComponentManagerImpl.java:259)
	at com.intellij.openapi.components.impl.ComponentManagerImpl.getPicoContainer(ComponentManagerImpl.java:58)
	at com.intellij.openapi.components.ServiceManager.doGetService(ServiceManager.java:47)
	at com.intellij.openapi.components.ServiceManager.getService(ServiceManager.java:42)
	at org.rust.cargo.project.settings.RustProjectSettingsServiceKt.getRustSettings(RustProjectSettingsService.kt:45)
	at org.rust.cargo.project.settings.RustProjectSettingsServiceKt.getToolchain(RustProjectSettingsService.kt:48)
	at org.rust.ide.annotator.RsCargoCheckAnnotator.checkProject(RsCargoCheckAnnotator.kt:199)
	at org.rust.ide.annotator.RsCargoCheckAnnotator$getCachedResult$1.compute(RsCargoCheckAnnotator.kt:55)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:49)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:222)
	at com.intellij.psi.impl.PsiCachedValue.getValueWithLock(PsiCachedValue.java:60)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:38)
	at org.rust.ide.annotator.RsCargoCheckAnnotator.doAnnotate(RsCargoCheckAnnotator.kt:66)
	at org.rust.ide.annotator.RsCargoCheckAnnotator.doAnnotate(RsCargoCheckAnnotator.kt:51)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.d(ExternalToolPass.java:224)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.access$500(ExternalToolPass.java:48)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.run(ExternalToolPass.java:172)
	at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:337)
	at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:327)
	at com.intellij.util.ui.update.MergingUpdateQueue.lambda$flush$1(MergingUpdateQueue.java:277)
	at com.intellij.util.ui.update.MergingUpdateQueue.flush(MergingUpdateQueue.java:291)
	at com.intellij.util.ui.update.MergingUpdateQueue.run(MergingUpdateQueue.java:246)
	at com.intellij.util.concurrency.QueueProcessor.runSafely(QueueProcessor.java:223)
	at com.intellij.util.Alarm$Request.runSafely(Alarm.java:418)
	at com.intellij.util.Alarm$Request.access$700(Alarm.java:343)
	at com.intellij.util.Alarm$Request$1.run(Alarm.java:385)
	at com.intellij.util.Alarm$Request.run(Alarm.java:396)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.intellij.util.concurrency.SchedulingWrapper$MyScheduledFutureTask.run(SchedulingWrapper.java:237)
	at com.intellij.util.concurrency.BoundedTaskExecutor$2.run(BoundedTaskExecutor.java:212)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

--- a/src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt
+++ b/src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt
@@ -3,8 +3,10 @@ package org.rust.ide.annotator
 import com.google.gson.JsonParser
 import com.intellij.lang.annotation.*
 import com.intellij.lang.annotation.Annotation
+import com.intellij.openapi.Disposable
 import com.intellij.openapi.application.Result
 import com.intellij.openapi.application.WriteAction
+import com.intellij.openapi.application.runReadAction
 import com.intellij.openapi.editor.Document
 import com.intellij.openapi.editor.Editor
 import com.intellij.openapi.fileEditor.FileDocumentManager
@@ -182,9 +184,16 @@ class RsCargoCheckAnnotator : ExternalAnnotator<CargoCheckAnnotationInfo, CargoC
 
     fun CargoSpan.isValid() = line_end > line_start || (line_end == line_start && column_end >= column_start)
 
+    // NB: executed asynchronously off EDT, so care must be taken not to access
+    // disposed objects
     fun checkProject(file: PsiFile): CargoCheckAnnotationResult? {
-        val module = file.module ?: return null
-        val projectRoot = module.cargoProjectRoot ?: return null
+        val (toolchain, projectPath, disposable: Disposable) = runReadAction {
+            val module = file.module ?: return@runReadAction null
+            if (module.isDisposed) return@runReadAction null
+            val projectRoot = module.cargoProjectRoot ?: return@runReadAction null
+            val toolchain = module.project.toolchain ?: return@runReadAction null
+            Triple(toolchain, projectRoot.path, module)
+        } ?: return null
 
         // We have to save the file to disk to give cargo a chance to check fresh file content.
         object : WriteAction<Unit>() {
@@ -196,8 +205,7 @@ class RsCargoCheckAnnotator : ExternalAnnotator<CargoCheckAnnotationInfo, CargoC
             }
         }.execute()
 
-        val output = module.project.toolchain?.cargo(projectRoot.path)?.checkProject(module)
-            ?: return null
+        val output = toolchain.cargo(projectPath).checkProject(disposable)
         if (output.isCancelled) return null
         return CargoCheckAnnotationResult(output.stdoutLines, file.project)
     }

